README.md
# spectrum
## [pub.dev Listing](https://pub.dev/packages/spectrum) | [API Doc](https://pub.dev/documentation/spectrum/latest) | [GitHub](https://github.com/Zabadam/spectrum)
##### Gradient API References:  [`GradientUtils`](https://pub.dev/documentation/spectrum/latest/gradients/GradientUtils.html) | [`GradientTween`](https://pub.dev/documentation/spectrum/latest/gradients/GradientTween-class.html) | [`Steps`](https://pub.dev/documentation/spectrum/latest/gradients/Steps-class.html) | [`FooShadedSteps`](https://pub.dev/documentation/spectrum/latest/gradients/RadialShadedSteps-class.html) | [`AnimatedGradient`](https://pub.dev/documentation/spectrum/latest/gradients/AnimatedGradient-class.html)
##### Color API References: [`Shading`](https://pub.dev/documentation/spectrum/latest/colors/Shading.html) | [`ColorOperators`](https://pub.dev/documentation/spectrum/latest/colors/ColorOperators.html) | [`ColorOperatorsMethods`](https://pub.dev/documentation/spectrum/latest/colors/ColorOperatorsMethods.html) | [`Spectrum`](https://pub.dev/documentation/spectrum/latest/colors/Spectrum-class.html) | [`SpectrumUtils`](https://pub.dev/documentation/spectrum/latest/colors/SpectrumUtils.html)
###### More:  [`ColorArithmetic` `Shades`](https://pub.dev/documentation/spectrum/latest/gradients/Shades-class.html) | [`StopsArithmetic` `Maths`](https://pub.dev/documentation/spectrum/latest/gradients/Maths-class.html) | [`SwatchMode`](https://pub.dev/documentation/spectrum/latest/colors/SwatchMode-class.html) | [`GradientStoryboard`](https://pub.dev/documentation/spectrum/latest/gradients/AnimatedGradient/storyboard.html) | [`NillGradients`](https://pub.dev/documentation/spectrum/latest/gradients/NillGradients.html) | [`MaterialColorToList`](https://pub.dev/documentation/spectrum/latest/colors/MaterialColorToList.html)
##### [🐸 Zaba.app ― simple packages, simple names.](#-zabaapp--simple-packages-simple-nameshttpspubdevpublisherszabaapppackages-other-flutter-packages-published-by-zabaapp)

<br />

[![spectrum header image](https://raw.githubusercontent.com/Zabadam/spectrum/main/doc/img/spectrum_50.gif)](https://raw.githubusercontent.com/Zabadam/spectrum/main/doc/img/spectrum.gif 'Click for full size')

A rainbow of `Color` and `Gradient` utilities such as [`GradientTween`](https://pub.dev/documentation/spectrum/latest/gradients/GradientTween-class.html),
gradient [`copyWith()`](https://pub.dev/documentation/spectrum/latest/gradients/GradientUtils/copyWith.html)
any potential parameters for the many types, [`Steps`-type gradients](#-steps),
[`complementary()`](https://pub.dev/documentation/spectrum/latest/colors/SpectrumUtils/complementary.html)
for colors, [`AnimatedGradient`](https://pub.dev/documentation/spectrum/latest/gradients/AnimatedGradient-class.html),
[`MaterialColor` generation](https://pub.dev/documentation/spectrum/latest/colors/SpectrumUtils/materialColor.html),
and more!

## 📚 Table of Contents

[![spectrum header](https://raw.githubusercontent.com/Zabadam/spectrum/main/doc/img/spectrum_15.gif)](#spectrum)

  1. 🌊 [Gradients](#-gradients 'Part I')
     1. 💫 [Interpolation](#-interpolation 'GradientTweens & bespoke IntermediateGradients')
     2. 🌈 [Steps](#-steps 'Hard-stepping "gradients" that intrinsically duplicate colors and stops lists')
        1. 🧹 [Softness](#-softness 'Controls the aliasing and may return Steps back into a Gradient')
        2. 🕶️ [Shaded Steps](#️-shaded-steps 'Evolves the concept of duplicating colors and stops to add color shading functionality')
     3. ⏰ [Animation](#-animation 'Feed an animation to an AnimatedGradient defined by a GradientStoryboard')
        1. 📖 [Gradient Storyboard](#-gradient-storyboard 'Lay out how the gradient will utilize its controller')
     4. 📋 [Override Copy With](#-override-copy-with 'Use this package with your own custom Gradient Types')
   
  2. 🎨 [Colors](#-colors 'Part II')
     1. ➗ [Operators](#-operators 'Although there is currently no division operator')
     2. 🗜️ [Spectrum](#️-spectrum 'Quick tools that would otherwise be floating functions, instead anchored to abstract class')
     3. 🤝 [Complementary Colors](#-complementary-colors 'Rapidly produce Lists of Colors with varying degrees of complement')
   
  3. 🛣️ [Roadmap](#️-roadmap 'The current TODO list')

<br />

## 🌊 Gradients

Feel free to consider any instantiated `Gradient` and use a fancy [`Gradient.copyWith()`](https://pub.dev/documentation/spectrum/latest/gradients/GradientUtils/copyWith.html)
method to override some or all of its properties. This method accepts all manner
of potential `Gradient` subtype properities and applies only the ones necessary
for the return.

> This [`GradientUtils`](https://pub.dev/documentation/spectrum/latest/gradients/GradientUtils.html)
> extension method is employed as the default "copy with" function throughout
> other portions of `spectrum`. This is achieved by a private method
> `spectrumCopyWith()` which fulfills the [`GradientCopyWith`](#-override-copy-with)
> type alias definition in default contexts where constructor argument
> `overrideCopyWith` is not initialized.

Another gradient utility is the simple getter `Gradient.reversed` which returns
a gradient with same properties, but whose `colors` (and interpretted or
explicit `stops`) are in reverse order.

### 💫 Interpolation

Smooth `Gradient` tweening not just by `Gradient.lerp()`... oh no,
that is not enough.

This package provides full bespoke [`IntermediateGradient`](https://pub.dev/documentation/spectrum/latest/gradients/IntermediateGradient-class.html)s
for truly smooth and eye-popping gradient animations between dissimilar gradient
types!

[![Incredibly beautiful GradientTweens with IntermediateGradients (accelerometer-translation by package:xl)](https://raw.githubusercontent.com/Zabadam/spectrum/main/doc/img/tween_small.gif)](https://raw.githubusercontent.com/Zabadam/spectrum/main/doc/img/tween.gif 'Click for full size | Incredibly beautiful GradientTweens with IntermediateGradients (accelerometer-translation by package:xl)')

The interpolated `Gradient` is obtained via [`GradientTween.evaluate(Animation animation)`](https://pub.dev/documentation/spectrum/latest/gradients/GradientTween-class.html)
or `transform(double t)`.

> 💡 \
> Until better `List` interpolation is implemented, feel free to experiment with
> `isAgressive`.

| ![transparent 600x1 pixel-pusher (ignore)](https://i.imgur.com/XCh0q2K.png) | [📚](#-table-of-contents 'Table of Contents') |
| --------------------------------------------------------------------------: | -------------------------------------------: |

### 🌈 Steps

Spectrum also adds entirely new gradient types called [`Steps`](https://pub.dev/documentation/spectrum/latest/gradients/Steps-class.html),
as well as shaded varieties named [`FooShadedSteps`](https://pub.dev/documentation/spectrum/latest/gradients/RadialShadedSteps-class.html).
Imagine a gradient that somewhat defeats the idea of a gradient and, instead of
smoothly transitioning between colors, creates a series of hard steps.

[![New Steps-type Gradients; three variety for Radial, Linear, and Sweep](https://raw.githubusercontent.com/Zabadam/spectrum/main/doc/img/steps.png)](https://raw.githubusercontent.com/Zabadam/spectrum/main/doc/img/steps.png 'Click for full size | New Steps-type Gradients; three variety for Radial, Linear, and Sweep')

This is achieved by intrinsically duplicating both the `List<Color>` of `colors`
as well as the (interpreted or explicit) `List<double>` of `stops`. As a product
of this process, any explicitly initalized list of `stops` ought to not end in
`1.0` lest that stop get eaten. For example, a three-color `Steps` could have 
`stops: [0.0, 0.3, 0.8]`.

| ![transparent 600x1 pixel-pusher (ignore)](https://i.imgur.com/XCh0q2K.png) | [📚](#-table-of-contents 'Table of Contents') |
| --------------------------------------------------------------------------: | -------------------------------------------: |

#### 🧹 Softness

A `softness` may be increased or zeroed out depending on preference or user
device display density. A high-DPI screen would look just fine with an aliased,
hard edge between colors. A low-DPI screen would benefit from a (still,
incredibly small) `softness` to provide as an additive for each second entry
when duplicating stops.

> Imagine `Steps.stops` is `[0.0, 0.3, 0.8]`. Providing a `softness` of `0.001`,
> the effective, resolved stops for this gradient is now:
> `[0.0, 0.001, 0.3, 0.3001, 0.8, 0.8001]`.

|                                                                                                                                                                                                                                                                                                                                                                                                         |                                                                                                                                                                                                   |
| :------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | :------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| [![Animation showing transition from Steps.softness == 0 to some greater value then transitioning back](https://raw.githubusercontent.com/Zabadam/spectrum/main/doc/img/softness_small.gif)](https://raw.githubusercontent.com/Zabadam/spectrum/main/doc/img/softness.gif 'Click for full size \| Animation showing transition from Steps.softness == 0 to some greater value then transitioning back') | A larger  `softness` has the effect of making `Steps` more like their original `Gradient` counterpart. <br /><br /> The neighboring example transitions from `softness: 0.0` -> `softness: 0.14`. |

| ![transparent 600x1 pixel-pusher (ignore)](https://i.imgur.com/XCh0q2K.png) | [📚](#-table-of-contents 'Table of Contents') |
| --------------------------------------------------------------------------: | -------------------------------------------: |

#### 🕶️ Shaded Steps

Now imagine *more* intrinsic color and stops math for a variety of
self-elaborating gradients... something like `FooShadedSteps` may be born.

|                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |
| :---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | :--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

CHANGELOG.md
# ![spectrum header image](https://raw.githubusercontent.com/Zabadam/spectrum/main/doc/img/spectrum_50.gif)
## **[0.2.2]** - 27 JUL 21
Corrects a [`NillGradients`](https://pub.dev/documentation/spectrum/latest/gradients/NillGradients.html)
bug where `Sweep`-styled nill gradients had both a `startAngle` and `endAngle`
of `0.0`, throwing an "optimized out" error.
- The end angle on these nill sweep gradients is now `0.01`.

## **[0.2.1+1]** - 26 JUL 21
Fixes README links and finishes documentation.

## **[0.2.1]** - 25 JUL 21
Modularizes package into `colors` & `gradients` mini-libraries.
- Import all-in-one as `import 'package:spectrum/spectrum.dart';`
- Or seperately such as `import 'package:spectrum/colors.dart';`

Reverts usage of Dart @ 2.13 non-function type aliases (AKA `typedef`s) as they
were throwing `dartdoc`.
- "`failed: Bad state: typedef GradientStoryboard = Map<GradientAnimation, dynamic> cannot have parameters`"

Simplifies name of `Color.generateComplements(5)` -> `Color.complementary(5)`.

## **[0.2.0]** - 23 JUL 21
Massive "fleshing out" of the project, establishing remaining goals.

- Added `FooShadedSteps`
- Added `AnimatedGradient`
  - See also: `GradientStoryboard`, `GradientAnimation`, `GradientProperty`,
  `TweenSpec`, & more
- Added `GradientCopyWith` template
  - Used to override the return type for the `Gradient.copyWith` method used
  from within either a `GradientTween` or `AnimatedGradient`
- Introduced `Gradient` getter `reversed` and convenience method `animate()`
- Introduced `isAgressive` flag for `GradientTween`
- For the generation of `MaterialColor`s via `Spectrum.materialColor`:
  - `Blend` -> `SwatchMode` with an additional value, `SwatchMode.complements`
  - Consider `Spectrum.materialAccent` to create a `MaterialAccentColor`
- Color complements are obtained by new `SpectrumUtils.generateComplements`
  - Which drives the `List<Color>` getter extensions on instantiated `Color`s,
  too; `Color.complementTriad`, for example

## **[0.1.0]** - 11 JUN 21
- Initial release.


example/lib/main.dart
import 'package:flutter/material.dart';

import 'package:foil/foil.dart';

// Modular:
import 'package:spectrum/colors.dart';
import 'package:spectrum/gradients.dart';
// All in One:
// import 'package:spectrum/spectrum.dart';

import 'palette.dart';

// ignore_for_file: public_member_api_docs

// const color = Color(0xFFBB3399);
// const color = Color(0xFFFF0000);
const color = Color(0xFFFF61ED);
// const color = Color(0xFF00AA33);
// const color = Color(0xFF3F82D3);
// const color = Color(0x72A46565);
// const color = Color(0xFFFC5C74);

void main() => runApp(const ExampleFrame());

/// [MaterialApp] frame.
class ExampleFrame extends StatelessWidget {
  /// [MaterialApp] frame.
  const ExampleFrame({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) => MaterialApp(
        debugShowCheckedModeBanner: false,
        home: const Example(),
        // home: const Palette(),
        themeMode: ThemeMode.light,
        theme: ThemeData(
          primarySwatch: Spectrum.materialColor(
            color.withOpacity(1),
            mode: SwatchMode.shade,
          ),
        ),
      );
}

/// Construct a [new Example] `Widget` to fill an [ExampleFrame].
class Example extends StatefulWidget {
  /// Fill an [ExampleFrame] with a [Scaffold] and [AppBar] whose body is a
  /// [PageView].
  ///
  /// Also constructs the `final gradient`s to provide to other demo objects.
  const Example({Key? key}) : super(key: key);

  @override
  _ExampleState createState() => _ExampleState();
}

class _ExampleState extends State<Example> {
  @override
  Widget build(BuildContext context) {
    // final gradient = LinearGradient(
    final gradient = RadialGradient(
      // final gradient = SweepGradient(
      // final gradient = LinearSteps(
      // final gradient = RadialSteps(
      // final gradient = SweepSteps(
      // final gradient = LinearShadedSteps(
      // final gradient = RadialShadedSteps(
      // final gradient = SweepShadedSteps(
      //
      center: const Alignment(0.5, -0.75),
      // tileMode: TileMode.decal,
      // stops: const [0.15, 0.25, 0.3, 0.75, 0.85],
      // colors: color.asMaterialColor.asList,
      colors: Spectrum.materialColor(color, mode: SwatchMode.shade, factor: 450)
          .asList,
      // colors: color.complementary(5),
      // colors: color.complementary(18),
      // colors: color.complementary(36),
    );

    final gradient2 = SweepShadedSteps(
      center: const Alignment(-0.3, -1.25),
      colors: color.complementary(18),
      shadeFunction: Shades.withAlpha,
      // softness: 0.005,
      // shadeFactor: 175,
      // shadeFactor: -150,
      // shadeFactor: 125,
      shadeFactor: 0.95,
    );

    final radialSteps = RadialSteps(colors: color.complementary(18));
    final linearSteps =
        LinearSteps(colors: (const Color(0xFF4BFF6F)).complementary(5));
    final sweepSteps =
        SweepSteps(colors: (const Color(0xFF71F2F0)).complementary(4));

    return Scaffold(
      backgroundColor: Colors.black,
      appBar: AppBar(title: const Text('spectrum')),
      body: PageView(
        physics: const BouncingScrollPhysics(),
        children: [
          // /// Fast shading
          // Container(
          //   decoration: BoxDecoration(
          //     gradient: RadialShadedSteps(
          //       colors: color.asMaterialColor.asList, // fast shading
          //       radius: 1.0,
          //     ),
          //   ),
          // ),

          /// Header "spectrum"
          const Header(),

          /// Full-page `Foil`-based `GradientTween` when tapped.
          /// See [Foil] to understand why there should at least be a black
          /// rectangle (`Container`) onto which to mask the `Foil`'s gradient.
          TestFoilTween(
            gradient,
            gradient2,
            isAgressive: true,
            isSlow: false,
            child: Container(color: Colors.black),
          ),
          TestFoilTween(
            gradient,
            gradient2,
            isAgressive: true,
            isSlow: true,
            child: Container(color: Colors.black),
          ),

          /// Same as above, one fast and one slow, but flag `isAgressive` is disabled. //
          // TestFoilTween(gradient0, gradient1, isAgressive: false, isSlow: false, child: Container(color: Colors.black),), //
          // TestFoilTween(gradient0, gradient1, isAgressive: false, isSlow: true, child: Container(color: Colors.black),), //

          /// Steps.softness demonstration
          Column(
            children: <Widget>[
              Flexible(
                child: AnimatedFoilDemo(
                  radialSteps,
                  isSlow: false,
                  isAgressive: true,
                  storyboard: _defaultStoryboard,
                ),
              ),
              Flexible(
                child: AnimatedFoilDemo(
                  linearSteps,
                  isSlow: false,
                  isAgressive: true,
                  storyboard: _defaultStoryboard,
                ),
              ),
              Flexible(
                child: AnimatedFoilDemo(
                  sweepSteps,
                  isSlow: false,
                  isAgressive: true,
                  storyboard: _defaultStoryboard,
                ),
              ),
            ],
          ),

          /// FooShadedSteps demonstration
          AnimatedFoilDemo(
            LinearShadedSteps(
              colors: (const Color(0xFF4BFF6F)).complementary(4),
              shadeFunction: Shades.withWhite, // default
            ),
            isSlow: false,
            isAgressive: true,
            storyboard: {
              /// "tweenSpec" expects a [TweenSpec], which itself is a
              /// `Map<GradientProperty, Tween<dynamic>>`.
              GradientAnimation.tweenSpec: {
                GradientProperty.shadeFactor: Tween<double>(begin: 0, end: 120),
                GradientProperty.distance: Tween<double>(begin: 0.6, end: 0.4),
              },
            },
          ),

          /// A column of color squares demonstrating the
          /// `Color.complementary()` method.
          const Palette(),

          /// A column of color squares demonstrating the
          /// `Color.complementary()` method.
          ///
          /// Also consider getters such as `Color.complementPair` or
          /// `Color.complementTetrad`.
          SingleChildScrollView(
              child: Column(children: const [
            SizedBox(height: 25),
            GenerateComplements(color, 1),
            GenerateComplements(color, 2),
            GenerateComplements(color, 3),
            SizedBox(height: 25),
            GenerateComplements(color, 4),
            GenerateComplements(color, 5),
            GenerateComplements(color, 6),
            SizedBox(height: 25),
            GenerateComplements(color, 36),
          ])),

          /// A `GradientTween` frozen in time for granular keyframe inspection.
          // Lerping(gradient, unwrappedGradient),
        ],
      ),
    );
  }
}

Animatable<Color?> colors = TweenSequence<Color?>([
  TweenSequenceItem<Color?>(
    weight: 1.0,
    tween: ColorTween(begin: Colors.red, end: Colors.blue),
    // tween: Tween<Color>(begin: Colors.red, end: Colors.blue),
  ),
  TweenSequenceItem<Color?>(
    weight: 1.0,
    tween: ColorTween(begin: Colors.blue, end: Colors.green),
    // tween: Tween<Color>(begin: Colors.blue, end: Colors.green),
  ),
  TweenSequenceItem<Color?>(
    weight: 1.0,
    tween: ColorTween(begin: Colors.green, end: Colors.yellow),
    // tween: Tween<Color>(begin: Colors.green, end: Colors.yellow),
  ),
  TweenSequenceItem<Color?>(
    weight: 1.0,
    tween: ColorTween(begin: Colors.lightBlue, end: Colors.purple),
    // tween: Tween<Color>(begin: Colors.blue, end: Colors.green),
  ),
  TweenSequenceItem<Color?>(
    weight: 1.0,
    tween: ColorTween(begin: Colors.purple, end: Colors.amber),
    // tween: Tween<Color>(begin: Colors.green, end: Colors.yellow),
  ),
  TweenSequenceItem<Color?>(
    weight: 1.0,
    tween: ColorTween(begin: Colors.amber, end: Colors.red),
    // tween: Tween<Color>(begin: Colors.yellow, end: Colors.red),
  ),
]);

class Header extends StatefulWidget {
  const Header({Key? key}) : super(key: key);

  @override
  _HeaderState createState() => _HeaderState();
}

class _HeaderState extends State<Header> with SingleTickerProviderStateMixin {
  late AnimationController _controller;
  late Animation<Color?> _color;

  @override
  void initState() {
    super.initState();
    _controller = AnimationController(
      duration: const Duration(seconds: 5),
      vsync: this,
    )
      ..addListener(() => setState(() {}))
      ..repeat(reverse: true);

    _color = colors.animate(_controller);
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return FittedBox(
      child: ShaderMask(
        // shaderCallback: (bounds) => GradientTween(
        //   begin: RadialShadedSteps(
        //     // colors: _color.value!.complementary(10),
        //     colors: _color.value!.asMaterialColor.asList,
        //     radius: 4,
        //     shadeFactor: 200,
        //   ),
        //   end: LinearGradient(
        //     colors: _color.value!.withWhite(50).complementary(6),
        //   ),
        //   isAgressive: false,
        // ).evaluate(_controller)!.createShader(bounds),
        shaderCallback: (bounds) => AnimatedGradient(
          controller: _controller,
          gradient: GradientTween(
            begin: RadialShadedSteps(
              // colors: _color.value!.complementary(10),
              colors: _color.value!.asMaterialColor.asList,
              radius: 4,
              shadeFactor: 200,
            ),
            end: LinearGradient(
              colors: _color.value!.withWhite(50).complementary(6),
            ),
            isAgressive: false,
          ).evaluate(_controller)!,
          storyboard: {
            // GradientAnimation.colorArithmetic: Shades.withOpacity,
            // GradientAnimation.stopsArithmetic: Maths.addition,
            GradientAnimation.tweenSpec: {
              GradientProperty.center: Tween<AlignmentGeometry?>(
                begin: const Alignment(-0.75, -0.75),
                end: const Alignment(5.75, 5.75),
              ),
              // GradientProperty.distance: Tween<double>(
              //   begin: 0,
              //   end: 1,
              // ),
            }
          },
        ).observe.createShader(bounds),
        child: const Text(
          ' spectrum ',
          style: TextStyle(
            fontSize: 400,
            fontWeight: FontWeight.w900,
            color: Colors.grey,
          ),
        ),
      ),
    );
  }
}

final _defaultStoryboard = {
  // GradientAnimation.none: null,

  /// "colorArithmetic" expects a [ColorArithmetic].
  // GradientAnimation.colorArithmetic: Shades.withOpacity,

  /// "stopsArithmetic" expects a [StopsArithmetic].
  // GradientAnimation.stopsArithmetic: Maths.subtraction,
  // GradientAnimation.stopsArithmetic:Maths.addition,
  // GradientAnimation.stopsArithmetic:Maths.division,

  /// "tweenSpec" expects a [TweenSpec], which itself is a
  /// `Map<GradientProperty, Tween<dynamic>>`.
  GradientAnimation.tweenSpec: {
    // GradientProperty.center: Tween<AlignmentGeometry>(
    //     begin: const Alignment(1, 1),
    //     end: const Alignment(-1, -1)),
    // GradientProperty.focal: Tween<AlignmentGeometry>(
    //     begin: const Alignment(3, -1),
    //     end: const Alignment(-3, -1)),
    // GradientProperty.radius:
    //Tween<double>(begin: 0.5, end: 0),
    // GradientProperty.focalRadius:
    // Tween<double>(begin:2,end:0),
    // // GradientProperty.startAngle: Tween<double>(
    // // begin: -1.0 * 3.1415927, end: 0.0 * 3.1415927),
    // // GradientProperty.endAngle:
    // // Tween<double>(begin: 2.0 * 3.1415927,end: 4.0 * 3.1415927),
    // GradientProperty.shadeFactor:
    //StepTween(begin: -200,end:0),
    GradientProperty.softness: Tween<double>(begin: 0.0, end: 0.14),
  },
};

class AnimatedFoilDemo extends StatefulWidget {
  /// Tapping this `Foil` widget (that renders the gradient) will
  /// tween the gradient to `gradient.reversed`.
  const AnimatedFoilDemo(
    this.gradient, {
    Key? key,
    this.storyboard,
    required this.isSlow,
    required this.isAgressive,
    this.duration = const Duration(seconds: 2),
    this.interval = const Interval(0.3, 1.0, curve: Curves.fastOutSlowIn),
    this.child,
  }) : super(key: key);

  final Gradient gradient;
  // final GradientStoryboard? storyboard;
  final Map<GradientAnimation, dynamic>? storyboard;
  final Duration duration;
  final Interval interval;
  final bool isSlow, isAgressive;
  final Widget? child;

  @override
  _AnimatedFoilDemoState createState() => _AnimatedFoilDemoState();
}

class _AnimatedFoilDemoState extends State<AnimatedFoilDemo>
    with SingleTickerProviderStateMixin {
  late AnimationController _controller;
  bool isUnwrapped = false;

  @override
  void initState() {
    super.initState();
    _controller = AnimationController(duration: widget.duration, vsync: this)
      ..addListener(() => setState(() {}))
      ..addStatusListener((status) {
        if (status == AnimationStatus.completed) _controller.reverse();
        if (status == AnimationStatus.dismissed) _controller.forward();
      })
      ..forward();
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final controller = CurvedAnimation(
      // Default duration is 2 seconds
      parent: _controller,
      // Default interval will hold state at initial position for
      // 30% of the duration as `Interval(0.3, 1.0, curve...)`
      curve: widget.interval,
    );

    return Foil(
      isUnwrapped: isUnwrapped,
      isAgressive: widget.isAgressive,
      useSensor: false,
      duration: widget.isSlow
          ? const Duration(milliseconds: 3000)
          : const Duration(milliseconds: 1500),
      curve: widget.isSlow ? Curves.linear : Curves.elasticOut,
      speed: widget.isSlow
          ? const Duration(milliseconds: 200)
          : const Duration(milliseconds: 1500),
      gradient: widget.gradient.animate(
        controller: controller,

        /// A [GradientStoryboard] is a
        /// `Map<GradientAnimation, dynamic>` where dynamic values are
        /// description objects that are type-checked according to
        /// their key.
        storyboard: widget.storyboard ?? _defaultStoryboard,
      ),
      unwrappedGradient:
          widget.gradient.reversed.copyWith(radius: 0.025).animate(
                controller: controller,
                storyboard: widget.storyboard ?? _defaultStoryboard,
              ),
      child: GestureDetector(
        onTap: () => setState(() => isUnwrapped = !isUnwrapped),
        child: Center(
          child: widget.child ??
              Container(color: Colors.white), // White for Foil masking
        ),
      ),
    );
  }
}

class TestFoilTween extends StatefulWidget {
  const TestFoilTween(
    this.gradient,
    this.unwrappedGradient, {
    Key? key,
    required this.isSlow,
    required this.isAgressive,
    this.child,
  }) : super(key: key);

  final Gradient gradient, unwrappedGradient;
  final bool isSlow, isAgressive;
  final Widget? child;

  @override
  _TestFoilTweenState createState() => _TestFoilTweenState();
}

class _TestFoilTweenState extends State<TestFoilTween> {
  bool isUnwrapped = false;

  @override
  Widget build(BuildContext context) {
    var foil = Foil(
      isUnwrapped: isUnwrapped,
      isAgressive: widget.isAgressive,
      gradient: widget.gradient,
      unwrappedGradient: widget.unwrappedGradient,
      blendMode: BlendMode.plus,
      speed: widget.isSlow
          ? const Duration(milliseconds: 200)
          : const Duration(milliseconds: 1500),
      duration: widget.isSlow
          ? const Duration(milliseconds: 3000)
          : const Duration(milliseconds: 1500),
      curve: widget.isSlow ? Curves.linear : Curves.elasticOut,
      child: widget.child ??
          const DecoratedBox(
            decoration: BoxDecoration(color: Colors.white),
          ),
    );

    return SizedBox(
      width: MediaQuery.of(context).size.width,
      height: MediaQuery.of(context).size.height,
      child: GestureDetector(
        onTap: () => setState(() => isUnwrapped = !isUnwrapped),
        child: Stack(fit: StackFit.expand, children: [foil]),
      ),
    );
  }
}

class Lerping extends StatelessWidget {
  const Lerping(this.gradient1, this.gradient2, {Key? key}) : super(key: key);

  final Gradient gradient1, gradient2;

  @override
  Widget build(BuildContext context) {
    // final test = gradient;
    // final test = gradient.asGradient;
    // final test = unwrappedGradient.asGradient;

    const t = 0.0;
    // ignore: unused_local_variable
    final interpolated =
        PrimitiveGradient.fromStretchLerp(gradient1, gradient2, t);

    // ignore: unused_local_variable
    final intermediate = IntermediateGradient(
      PrimitiveGradient.fromStretchLerp(gradient1, gradient2, t),
      // PrimitiveGradient.interpolateFrom(gradient, unwrappedGradient, 1.0),
      GradientPacket(gradient1, gradient2, t),
      // GradientPacket(gradient, unwrappedGradient, 1.0),
      // PrimitiveGradient.interpolateFrom(test, unwrappedGradient, 0.0),
      // GradientPacket(test, unwrappedGradient, 0.0),
    );

    final tween = GradientTween(
      begin: gradient1,
      // begin: intermediate,
      // begin: test,
      // end: const LinearGradient(colors: [Colors.green, Colors.blue]),
      end: gradient2,
      // isAgressive: false,
    );

    return Column(
      mainAxisAlignment: MainAxisAlignment.center,
      children: [
        // Center(
        //   child: Container(
        //     width: 1460,
        //     height: 825 / 5,
        //     decoration: BoxDecoration(gradient: gradient),
        //   ),
        // ),
        Center(
          child: Container(
            width: 1460,
            height: 825 / 3,
            // height: 825 / 5,
            decoration: BoxDecoration(gradient: gradient1),
          ),
        ),
        Center(
          child: Container(
            width: 1460,
            height: 825 / 3,
            // height: 825 / 5,
            // decoration:
            //   BoxDecoration(gradient: unwrappedGradient.asGradient),
            decoration: BoxDecoration(gradient: gradient2),
          ),
        ),
        Center(
          child: Container(
            width: 1460,
            height: 825 / 3,
            // height: 825 / 5,
            decoration: BoxDecoration(gradient: tween.lerp(0)),
          ),
        ),
        // Center(
        //   child: Container(
        //     width: 1460,
        //     height: 825 / 5,
        //     decoration: BoxDecoration(gradient: tween.lerp(0.0)),
        //     // decoration: BoxDecoration(
        //     //     gradient: Gradient.lerp(gradient, unwrappedGradient, 0)),
        //     // gradient.asGradient, unwrappedGradient.asGradient, 0)),
        //   ),
        // ),
      ],
    );
  }
}
